src/web_socket.toit

```diff
@@ -0,0 +1,405 @@
+      return nonce
+    response_writer.write_headers STATUS_BAD_REQUEST --message=message
+    return null
+
+  static response_ nonce/string -> string:
+    expected_response := base64.encode
+        sha1 nonce + "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"
+    return expected_response
+
+/**
+A writer for writing a single message on a WebSocket connection.
+*/
+class WebSocketWriter:
+  owner_ /WebSocket? := ?
+  size_ /int?
+  remaining_in_fragment_ /int := 0
+  fragment_sent_ /bool := false
+
+  constructor.private_ .owner_ .size_:
+
+  write data from=0 to=data.size -> int:
+    if owner_ == null: throw "ALREADY_CLOSED"
+    total_size := to - from
+    while true:
+      while from < to and remaining_in_fragment_ != 0:
+        size := min to - from remaining_in_fragment_
+        // We don't use slices because data might be a string with UTF-8
+        // sequences in it.
+        owner_.write_ data from (from + size)
+        from += size
+        remaining_in_fragment_ -= size
+      if from == to: return total_size
+
+      opcode := data is string ? OPCODE_TEXT_ : OPCODE_BINARY_
+      if fragment_sent_:
+        opcode = OPCODE_CONTINUATION_
+      else:
+        fragment_sent_ = true
+
+      header := ?
+      if size_:
+        // We know the size.  Write a single fragment with the fin flag and the
+        // exact size.
+        if opcode == OPCODE_CONTINUATION_: throw "TOO_MUCH_WRITTEN"
+        remaining_in_fragment_ = size_
+        if size_ > 0xffff:
+          header = ByteArray 10
+          header[1] = EIGHT_BYTE_SIZE_
+          BIG_ENDIAN.put_int64 header 2 size_
+        else if size_ > MAX_ONE_BYTE_SIZE_:
+          header = ByteArray 4
+          header[1] = TWO_BYTE_SIZE_
+          BIG_ENDIAN.put_uint16 header 2 size_
+        else:
+          header = ByteArray 2
+          header[1] = size_
+        header[0] = opcode | FIN_FLAG_
+      else:
+        // We don't know the size.  Write multiple fragments of up to
+        // 126 bytes.
+        remaining_in_fragment_ = min MAX_ONE_BYTE_SIZE_ data.size - from
+        header = ByteArray 2
+        header[0] = opcode
+        header[1] = remaining_in_fragment_
+      owner_.write_ header
+
+  close:
+    if remaining_in_fragment_ != 0: throw "TOO_LITTLE_WRITTEN"
+    if owner_:
+      if size_ == null:
+        header := ByteArray 2
+        header[0] = FIN_FLAG_ | (fragment_sent_ ? OPCODE_CONTINUATION_ : OPCODE_BINARY_)
+        owner_.write_ header
+      owner_.writer_close_ this
+      owner_ = null
+
+/**
+A reader for an individual message sent to us.
+*/
+class WebSocketReader implements reader.Reader:
+  owner_ /WebSocket? := ?
+  is_text /bool
+  fragment_reader_ /FragmentReader_ := ?
+
+  /**
+  If the size of the incoming message is known, then it is returned, otherwise
+    null is returned.
+  Since a normal size method cannot return null, the WebSocketReader does not
+    implement $reader.SizedReader.
```

Comment:
You could create a subclass that implements SizedReader (and just overwrites the `size` field to return `-> int`).
An `is SizedReader` would then work.

Really not sure it's worth it.